README.md
# homebridge-screenlogic

[![NPM version](https://img.shields.io/npm/v/homebridge-screenlogic)](https://npmjs.org/package/homebridge-screenlogic)
![License](https://img.shields.io/npm/l/homebridge-screenlogic)
[![Downloads](https://img.shields.io/npm/dm/homebridge-screenlogic.svg)](https://npmjs.org/package/homebridge-screenlogic)

## Screen Logic plug-in for [Homebridge](https://github.com/nfarina/homebridge) using the [node-screenlogic](https://github.com/parnic/node-screenlogic) library.

<img src="https://user-images.githubusercontent.com/249172/86522408-04f3d980-be12-11ea-8fc2-e85904a8b415.png" width="300"> <img src="https://user-images.githubusercontent.com/249172/86522409-0a512400-be12-11ea-8c9f-4165e9fe76a5.png" width="300">

## Requirements

This plugin was recently rewritten in Typescript to use the latest capabilities of Homebridge 1.x, so it requires at a minimum:

1. Homebridge >= 1.0.0
2. Node >= 10.17.0

## Installation

<!-- 2. Clone (or pull) this repository from github into the same path Homebridge lives (usually `/usr/local/lib/node_modules`). Note: the code currently on GitHub is in beta, and is newer than the latest published version of this package on `npm` -->

1. Install homebridge using: `npm install -g homebridge`
2. Install this plug-in using: `npm install -g homebridge-screenlogic`
3. Update your configuration file. See example `config.json` snippet below.

## Configuration

Configuration samples (edit `~/.homebridge/config.json`):

### UDP Broadcast

This performs a UDP broadcast on 255.255.255.255, port 1444, so ensure your network supports UDP broadcasts and the device is on the same subnet.

```json
"platforms": [
        {
            "platform": "ScreenLogic"
        }
    ],
```

### Direct connection via IP Address

Use this when you know the local static IP address.

```json
"platforms": [
        {
            "platform": "ScreenLogic",
            "ip_address": "192.168.0.100"
        }
    ],
```

- `"port"` is optional and defaults to 80
- `"username"` is optional, but is recommended as it is used to keep accessory UUIDs consistent. Should be in the format `"Pentair: XX-XX-XX"`.

### Remote connection via gateway

Use this to go through Pentair servers.

```json
"platforms": [
        {
            "platform": "ScreenLogic",
            "username": "Pentair: XX-XX-XX",
            "password": "..."
        }
    ],
```

## Optional fields:

- `"hidden_circuits"` comma-separated list of circuit names to hide. Set this for circuits you don't want showing up as switches. (ie.., `"Aux 6,Floor Cleaner"`).

- `"hideAirTemperatureSensor"` hides the air temperature sensor. Default is `false`.

- `"hidePoolTemperatureSensor"` hides the pool temperature sensor, which is redundant if you are showing pool thermostat. Default is `false`.

- `"hideSpaTemperatureSensor"` hides the spa temperature sensor, which is redundant if you are showing spa thermostat. Default is `false`.

- `"hidePoolThermostat"` hides the pool thermostat (aka, pool heater) if you don't want to allow changes via HomeKit. Default is `false`.

- `"hideSpaThermostat"` hides the spa thermostat (aka, spa heater) if you don't want to allow changes via HomeKit. Default is `false`.

- `"statusPollingSeconds"` time in seconds to poll for pool statu. Default is 60 seconds.

- `createLightColorSwitches` will create a "Pool Lights" accessory that contains switches for turing on light modes/colors. Default is `false`.

- `disabledLightColors` an array of strings with the names of light mode/colors you want to _disable_. Default is an empty array.

## Sample config

```json
{
  "platform": "ScreenLogic",
  "hidePoolTemperatureSensor": false,
  "hideSpaTemperatureSensor": true,
  "hideAirTemperatureSensor": true,
  "hidePoolThermostat": false,
  "hideSpaThermostat": false,
  "statusPollingSeconds": 60,
  "createLightColorSwitches": true,
  "disabledLightColors": ["Pool Mode Party", "Pool Mode Romance"],
  "hidden_circuits": "Fountains,Floor Cleaner,Aux 6"
}
```

# Implemented HomeKit Accessory Types

## Air Temperature

- _TemperatureSensor_ accessory (Air) indicating the ambient temperature where thee screenlogic hardware is located

## Pool

- _TemperatureSensor_ accessory (Pool) indicating the ambient temperature of the pool (last known temperature if pool isn't running)

## Spa

- _TemperatureSensor_ accessory (Spa) indicating the ambient temperature of the Spa (last known temperature if pool isn't running)

## Circuits

- creates a _Switch_ accessory for each discovered circuit (i.e., Pool, Spa, Jets, Pool Light, Spa Light, etc)

## Pool Heater

- _Thermostat_ accessory with ambient temperature, mode control (heat/cool/auto/off), and target temperature control. See also [Note on Pool/Spa Heater](#note-on-poolspa-heater) below.

## Spa Heater

- _Thermostat_ accessory with ambient temperature, mode control (heat/cool/auto/off), and target temperature control. See also [Note on Pool/Spa Heater](#note-on-poolspa-heater) below.

## Light Colors

- _Light Colors_ accessory with multiple switches for setting a light mode/color. See also [Note on Light Colors](#note-on-light-colors)

# Note on Pool/Spa Heater

The Pool and Spa Heater accessories are exposed as Thermostats in HomeKit. Since the semantics are slightly different between Pentair heat mode and a thermostat target heating state, a mapping is required.

## Mapping

I picked the following mapping, which seemed like the most logical mapping:

| Pentair Heat Mode | Thermostat State |
| ----------------- | ---------------- |
| Off               | Off              |
| Heater            | Heat             |
| Solar Preferred   | Auto             |
| Solar             | Cool             |

The only strange one is mapping `Solar` to `Cool`. I decided to go that route given the first three were fairly obvious (to me at least).

An alternative would be to expose three distinct on/off switches that represent each mode, and then ignore state changes (and maybe just allowing Off/Heat).

## "On" State

The other compromise is that the pool and spa heaters do _not_ turn the pool and/or spa on or off, they just change the heat mode.

i.e., if you want to heat the spa, you need to do two things:

1. turn on the Spa (via the Spa Switch)
2. make sure the Spa Heater is set to something other than off (most likely Heat)

This should work well in practice though, as you will generally have a set target temperature and mode, and then just turn the spa on/off without mucking with the thermostat.

This also means that even if the Pool/Spa is turned off and you open the Pool/Spa Heater it mght say "HEATING TO". It will not actually being heating unless the corresponding Pool/Spa switch is turned on.

# Note on Light Colors

The light color commands (for modes and colors) are exposed as switches in HomeKit if you enable `createLightColorSwitches` (it is `false` by default).

## Semantics

Since I don't have the ability to query the current state of which color/mode is active (Screenlogic app doesn't show it either), I implemented the following behavior:

1. light mode/color switch is turned on (i.e., Pool Mode Sunset)
1. light command is sent to Screenlogic controller
1. after a few seconds, the same switch is turned off to give feedback that the command was sent
1. all pool/spa lights will be turned on (if they aren't already), and set to that mode/color. This is done by the screen logic controller itself.

## Single Tile

Instead of cluttering the room with a bunch of switches, they are all shown in HomeKit as a single accessory called "Light Colors". Tap on the tile to expand, and then turn on the individual switch for the desired mode/color.

If you'd like to show them as separate tiles, you can tap the gear icon at the bottom of the expanded tile (or slide up at the bottom) and then select "Show as Separate Tiles". If you are showing as separate tiles and want to revert back to a single tile, you can select any switch and then select "Show as Single Tile"

## Renaming

While showing the light switches as separate tiles, you can rename them if desired, which will let you pick a different name to use with Siri. After renaming you can then show as a single tile again if you'd like.

## Siri

You can set a mode/color by saying the mode/color name:

- "Hey Siri, turn on pool mode sunset"
- "Hey Siri, turn on pool color blue"

Which will set the selected mode/color and turn on all the lights if they aren't on.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## V2.2.1 - 2020-11-9

### Fixed

- set min/max on `CurrentTemperature` characteristic to allow temperatures below 0 degrees Celsius. Fixes "unavailable" error message from HomeKit went temperatures dropped below 0.

## V2.2.0 - 2020-9-26

### Fixed

- fix light colors

- fixes for new TypesSript errors

## V2.1.1 - 2020-7-5

### Added

- DEBUG: add additional debug output when successfully sending commands

- DEBUG: dump out raw pool `controllerConfig` message from `node-screenlogic`

## V2.1.0 - 2020-7-4

### Added

- add `createLightColorSwitches` config that creates switches for setting pool light modes/colors. See [README.md](./README.md) for more details
- add `disabledLightColors` config for disabling light modes/colors that you don't want to create switches for. See [README.md](./README.md) for more details

### Fixed

- add catch handlers for controller promises to log errors

## V2.0.3 - 2020-6-27

### Fixed

- simplified the way `get` triggered a refresh
- Typescript cleanup

## V2.0.2 - 2020-6-22

### Fixed

- restore Hoobs support by using `api.registerPlatform(PLUGIN_NAME, PLATFORM_NAME, ScreenLogicPlatform)`

## V2.0.1 - 2020-6-22

### Added

- add `hideAirTemperatureSensor` config that hides the air temperature sensor. Default value is `false`.

- don’t use status faults to indicate polling error

## V2.0.0 - 2020-6-21 (MAJOR RELEASE)

### Added

- rewrite in Typescript to provide much cleaner/safer base
- minimum of Homebridge 1.1.1 is required
- support DynamicPlatformPlugin
- properly add/remove/update cached accessories on startup

### Fixed

- caching/persisting of accessories, so if bridge is unavailable when server starts the previous configured shades will remain

- add retry logic for obtaining the initial config from the controller if there is an error contacting it

- try to cleanup the way I was setting/getting/updating of service characteristics to be more proper

## v1.4.2 - 2020-5-29

### Fixed

- fixes #4 by adding extra checks to see if spa or pool circuits exist

## v1.4.1 - 2020-5-25

### Fixed

- updated to `node-screenlogic` v1.4.0 which emits errors to the caller

- add `.on('error')` handlers to propagate errors back to promises via reject

## v1.4.0 - 2020-4-18

### Fixed

- since we are using a shared Promise, make sure to have a `.catch` handler when also setting up the `.finally`, otherwise we'll get an unhandled promise rejection error

## v1.3.5 - 2020-3-20

### Fixed

- fixes #3 by clamping set point so it is just above/below the min/max set point

## v1.3.4 - 2020-3-20

### Added

- dump out pool config and status when debug is turned on

- add a default case for `mapHeatModeToTargetHeatingCoolingState`

## v1.3.3 - 2020-3-19

### Fixed

- attempt to fix issue #3

was incorrectly using `Characteristic.CurrentHeaterCoolerState.{OFF,HEAT}` instead of `Characteristic.CurrentHeatingCoolingState.{HEAT,OFF}`

before fixing I was seeing:

```
[3/19/2020, 2:46:56 PM][screenlogic] Pool Heater set heatingCoolingState undefined
[3/19/2020, 2:46:56 PM][screenlogic] Pool Heater set heatingCoolingState 0
```

afer fixing:

```
[3/19/2020, 2:49:29 PM][screenlogic] Pool Heater set heatingCoolingState 0
[3/19/2020, 2:49:29 PM][screenlogic] Pool Heater set heatingCoolingState 0
```

## v1.3.2 - 2020-3-14

### Added

- updated images in README.md

## v1.3.1 - 2020-3-14

### Added

- updated images in README.md

## v1.3.0 - 2020-3-14

### Added

- add "Pool Heater" and "Spa Heater" accessories which are exposed as HomeKit Thermostats. Fixes #1. See README.md for details

- add `hidePoolTemperatureSensor` config that hides the pool temperature sensor, which is redundant if you are showing pool thermostat

- add `hideSpaTemperatureSensor` config that hides the spa temperature sensor, which is redundant if you are showing spa thermostat

- add `hidePoolThermostat` config that hides the pool thermostat (aka, pool heater) if you don't want to allow changes via HomeKit

- add `hideSpaThermostat` config that hides the spa thermostat (aka, spa heater) if you don't want to allow changes via HomeKit

### Fixed

- fixes #2, typo in description

## v1.2.1 - 2020-1-4

### Fixed

- actually set default polling to 60 seconds

## v1.2.0 - 2020-1-4

### Added

- add background polling for status updates to enable automations to work

- add `statusPollingSeconds` option to config, defaults to 60 seconds

- expontential backoff on error

## v1.1.3 - 2020-1-3

### Fixed

- promise.then(f1,f2) -> promise.then(f1).catch(f2) cleanup

- remove bluebird dependency since not using it

## v1.1.2 - 2020-1-3

### Fixed

- log connected/errors in \_refreshStatus


src/platform.ts
import {
  API,
  DynamicPlatformPlugin,
  Logger,
  PlatformAccessory,
  PlatformConfig,
  Service,
  WithUUID,
  Characteristic,
  CharacteristicGetCallback,
} from 'homebridge'

import { PLATFORM_NAME, PLUGIN_NAME } from './settings'
import { TemperatureAccessory, TemperatureAccessoryContext } from './temperatureAccessory'
import { CircuitAccessory, CircuitAccessoryContext } from './circuitAccessory'
import { ThermostatAccessory, ThermostatAccessoryContext } from './thermostatAccessory'
import { SetColorAccessory, SetColorAccessoryContext } from './setColorAccessory'

import { Controller, PoolConfig, PoolStatus } from './controller'

type CharacteristicType = WithUUID<{ new (): Characteristic }>

export interface AccessoryAdaptor<T> {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  generateUUID(platform: ScreenLogicPlatform, context: Record<string, any>): string
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  sameContext(a: Record<string, any>, b: Record<string, any>): boolean
  factory(platform: ScreenLogicPlatform, accessory: PlatformAccessory): T
}

export interface ScreenLogicPlatformConfig extends PlatformConfig {
  ip_address?: string
  port?: number
  username?: string
  password?: string
  hidePoolTemperatureSensor: boolean
  hideSpaTemperatureSensor: boolean
  hideAirTemperatureSensor: boolean
  hidePoolThermostat: boolean
  hideSpaThermostat: boolean
  statusPollingSeconds: number
  createLightColorSwitches: boolean
  disabledLightColors: [string]
}

const POOL_TEMP_NAME = 'Pool'
const SPA_TEMP_NAME = 'Spa'
const AIR_TEMP_NAME = 'Air'
const POOL_THERMOSTAT_NAME = 'Pool Heater'
const SPA_THERMOSTAT_NAME = 'Spa Heater'

/**npm
 * HomebridgePlatform
 * This class is the main constructor for your plugin, this is where you should
 * parse the user config and discover/register accessories with Homebridge.
 */
export class ScreenLogicPlatform implements DynamicPlatformPlugin {
  public readonly Service: typeof Service = this.api.hap.Service
  public readonly Characteristic: typeof Characteristic = this.api.hap.Characteristic

  // this is used to track restored cached accessories
  private restoredAccessories: PlatformAccessory[] = []

  // bridge controller to talk to shades
  private controller: Controller

  // configuration
  public config: ScreenLogicPlatformConfig

  // fetched config
  private poolConfig?: PoolConfig

  // the time (from Date.now()) when we last initiated or did a refresh
  private lastRefreshTime = 0

  private poolTempAccessory?: TemperatureAccessory
  private spaTempAccessory?: TemperatureAccessory
  private airTempAccessory?: TemperatureAccessory

  private poolThermostatAccessory?: ThermostatAccessory
  private spaThermostatAccessory?: ThermostatAccessory

  private circuitAccessories: CircuitAccessory[] = []

  constructor(public readonly log: Logger, config: PlatformConfig, public readonly api: API) {
    this.log.debug('Finished initializing platform', PLATFORM_NAME)

    this.config = config as ScreenLogicPlatformConfig
    // do this first to make sure we have proper defaults moving forward
    this.applyConfigDefaults(config)

    this.controller = new Controller({
      log: this.log,
      ip_address: this.config.ip_address,
      port: this.config.port,
      username: this.config.username,
      password: this.config.password,
    })

    // When this event is fired it means Homebridge has restored all cached accessories from disk.
    // Dynamic Platform plugins should only register new accessories after this event was fired,
    // in order to ensure they weren't added to homebridge already. This event can also be used
    // to start discovery of new accessories.
    this.api.on('didFinishLaunching', () => {
      log.debug('Executed didFinishLaunching callback')
      // run the method to discover / register your devices as accessories
      this.discoverDevices(0)
    })
  }

  private discoverDevices(retryAttempt: number) {
    // test
    this.log.info('discoverDevices')

    const pollingInterval = 60 // TODO: get from config

    this.controller
      .getPoolConfig()
      .then(config => {
        this.poolConfig = config
        this.log.debug('got pool config', this.poolConfig)
        this.log.info(
          `discoverDevices connected: ${this.poolConfig.deviceId} ${this.poolConfig.softwareVersion}`,
        )
        this.setupDiscoveredAccessories(this.poolConfig)
      })
      .catch(err => {
        // on error, start another timeout with backoff
        const timeout = this.backoff(retryAttempt, pollingInterval)
        this.log.error(
          `discoverDevices retryAttempt: ${retryAttempt} timeout: ${timeout} error: ${err}`,
        )
        setTimeout(() => this.discoverDevices(retryAttempt + 1), timeout * 1000)
      })
  }

  private setupDiscoveredAccessories(poolConfig: PoolConfig) {
    // this is used to track active accessories uuids
    const activeAccessories = new Set<string>()

    // make pool temperature sensor if needed
    if (poolConfig.hasPool && !this.config.hidePoolTemperatureSensor) {
      this.poolTempAccessory = this.configureAccessoryType(TemperatureAccessory.makeAdaptor(), {
        displayName: POOL_TEMP_NAME,
        type: POOL_TEMP_NAME,
      } as TemperatureAccessoryContext)
      activeAccessories.add(this.poolTempAccessory.UUID)
    }

    // make spa temperature sensor if needed
    if (poolConfig.hasSpa && !this.config.hideSpaTemperatureSensor) {
      this.spaTempAccessory = this.configureAccessoryType(TemperatureAccessory.makeAdaptor(), {
        displayName: SPA_TEMP_NAME,
        type: SPA_TEMP_NAME,
      } as TemperatureAccessoryContext)
      activeAccessories.add(this.spaTempAccessory.UUID)
    }

    // make air temperature sensor if needed
    if (!this.config.hideAirTemperatureSensor) {
      this.airTempAccessory = this.configureAccessoryType(TemperatureAccessory.makeAdaptor(), {
        displayName: AIR_TEMP_NAME,
        type: AIR_TEMP_NAME,
      } as TemperatureAccessoryContext)
      activeAccessories.add(this.airTempAccessory.UUID)
    }

    // make pool thermostat if needed
    if (poolConfig.hasPool && !this.config.hidePoolThermostat) {
      this.poolThermostatAccessory = this.configureAccessoryType(
        ThermostatAccessory.makeAdaptor(),
        {
          displayName: POOL_THERMOSTAT_NAME,
          bodyType: 0,
          minSetPoint: this.normalizeTemperature(poolConfig.poolMinSetPoint),
          maxSetPoint: this.normalizeTemperature(poolConfig.poolMaxSetPoint),
        } as ThermostatAccessoryContext,
      )
      activeAccessories.add(this.poolThermostatAccessory.UUID)
    }

    // make spa thermostat if needed
    if (poolConfig.hasSpa && !this.config.hideSpaThermostat) {
      this.spaThermostatAccessory = this.configureAccessoryType(ThermostatAccessory.makeAdaptor(), {
        displayName: SPA_THERMOSTAT_NAME,
        bodyType: 1,
        minSetPoint: this.normalizeTemperature(poolConfig.spaMinSetPoint),
        maxSetPoint: this.normalizeTemperature(poolConfig.spaMaxSetPoint),
      } as ThermostatAccessoryContext)
      activeAccessories.add(this.spaThermostatAccessory.UUID)
    }

    // filter out hidden circuits
    const hiddenNames = (this.config.hidden_circuits as string) || ''
    const hiddenCircuits = new Set(hiddenNames.split(',').map(item => item.trim()))

    poolConfig.circuits = poolConfig.circuits.filter(circuit => {
      return !hiddenCircuits.has(circuit.name)
    })

    for (const circuit of poolConfig.circuits) {
      const accessory = this.configureAccessoryType(CircuitAccessory.makeAdaptor(), {
        displayName: circuit.name,
        id: circuit.id,
      } as CircuitAccessoryContext)
      this.circuitAccessories.push(accessory)
      activeAccessories.add(accessory.UUID)
    }

    if (this.config.createLightColorSwitches) {
      const accessory = this.configureAccessoryType(SetColorAccessory.makeAdaptor(), {
        displayName: 'Light Colors',
        id: 1,
      } as CircuitAccessoryContext)
      activeAccessories.add(accessory.UUID)
    }

    // unregister orphaned accessories
    const staleAccessories = this.restoredAccessories.filter(
      accessory => !activeAccessories.has(accessory.UUID),
    )

    if (staleAccessories.length) {
      const staleNames = staleAccessories.map(accessory => accessory.displayName)
      this.log.info('unregistering accessories', staleNames)
      this.api.unregisterPlatformAccessories(PLUGIN_NAME, PLATFORM_NAME, staleAccessories)
    }

    // start polling for status
    this.pollForStatus(0)
  }

  /**
   * This function is invoked when homebridge restores cached accessories from disk at startup.
   * It should be used to setup event handlers for characteristics and update respective values.
   */
  public configureAccessory(accessory: PlatformAccessory) {
    this.log.info('Loading accessory from cache:', accessory.displayName)

    // add the restored accessory to the accessories cache so we can track if it has already been registered
    this.restoredAccessories.push(accessory)
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private configureAccessoryType<T>(adaptor: AccessoryAdaptor<T>, context: Record<string, any>): T {
    // generate a unique id for this shade based on context
    const uuid = adaptor.generateUUID(this, context)

    // see if an accessory with the same uuid has already been registered and restored from
    // the cached devices we stored in the `configureAccessory` method above
    const existingAccessory = this.restoredAccessories.find(accessory => accessory.UUID === uuid)

    if (existingAccessory) {
      // the accessory already exists
      this.log.info('Restoring existing accessory from cache:', existingAccessory.displayName)

      // update the context if it has changed
      if (!adaptor.sameContext(context, existingAccessory.context)) {
        existingAccessory.context = context
        this.log.info('Updating existing accessory:', context.displayName)
        this.api.updatePlatformAccessories([existingAccessory])
      }
      // create the accessory handler for the restored accessory
      // this is imported from `platformAccessory.ts`
      return adaptor.factory(this, existingAccessory)
    } else {
      // the accessory does not yet exist, so we need to create it
      this.log.info('Adding new accessory:', context.displayName)

      // create a new accessory
      const accessory = new this.api.platformAccessory(context.displayName, uuid)

      // store a copy of the device object in the `accessory.context`
      accessory.context = context

      // create the accessory handler for the newly create accessory
      const newAccessory = adaptor.factory(this, accessory)

      // link the accessory to platform
      this.api.registerPlatformAccessories(PLUGIN_NAME, PLATFORM_NAME, [accessory])

      return newAccessory
    }
  }

  /** start polling process with truncated exponential backoff: https://cloud.google.com/storage/docs/exponential-backoff */
  private pollForStatus(retryAttempt: number) {
    const pollingInterval = this.config.statusPollingSeconds

    this.refreshStatus()
      .then(() => {
        // on success, start another timeout at normal pollingInterval
        this.log.debug('_pollForStatus success, retryAttempt:', retryAttempt)
        setTimeout(() => this.pollForStatus(0), pollingInterval * 1000)
      })
      .catch(err => {
        // on error, start another timeout with backoff
        const timeout = this.backoff(retryAttempt, pollingInterval)
        this.log.error('_pollForStatus retryAttempt:', retryAttempt, 'timeout:', timeout, err)
        setTimeout(() => this.pollForStatus(retryAttempt + 1), timeout * 1000)
      })
  }

  /** gets status,  updates accessories, and resolves */
  private async refreshStatus() {
    try {
      this.lastRefreshTime = Date.now()
      const poolStatus = await this.controller.getPoolStatus()
      this.log.debug('connected:', this.poolConfig?.deviceId, '(getStatus)')
      // update all values
      this.updateAccessories(poolStatus, undefined)
      return null
    } catch (err) {
      this.updateAccessories(undefined, err)
      throw err
    }
  }

  private updateAccessories(status?: PoolStatus, _err?: Error) {
    if (status) {
      this.airTempAccessory?.updateCurrentTemperature(
        this.normalizeTemperature(status.airTemperature),
      )
      this.airTempAccessory?.updateStatusActive(true)

      this.poolTempAccessory?.updateCurrentTemperature(
        this.normalizeTemperature(status.poolTemperature),
      )
      this.poolTempAccessory?.updateStatusActive(status.isPoolActive)

      this.spaTempAccessory?.updateCurrentTemperature(
        this.normalizeTemperature(status.spaTemperature),
      )
      this.spaTempAccessory?.updateStatusActive(status.isSpaActive)

      if (this.poolThermostatAccessory) {
        this.poolThermostatAccessory.updateCurrentTemperature(
          this.normalizeTemperature(status.poolTemperature),
        )
        this.poolThermostatAccessory.updateTargetTemperature(
          this.normalizeTemperature(status.poolSetPoint),
        )
        this.poolThermostatAccessory.updateCurrentHeatingCoolingState(
          status.isPoolHeating
            ? this.Characteristic.CurrentHeatingCoolingState.HEAT
            : this.Characteristic.CurrentHeatingCoolingState.OFF,
        )

        this.poolThermostatAccessory.updateTargetHeatingCoolingState(
          this.mapHeatModeToTargetHeatingCoolingState(status.poolHeatMode),
        )
      }

      if (this.spaThermostatAccessory) {
        this.spaThermostatAccessory.updateCurrentTemperature(
          this.normalizeTemperature(status.spaTemperature),
        )
        this.spaThermostatAccessory.updateTargetTemperature(
          this.normalizeTemperature(status.spaSetPoint),
        )
        this.spaThermostatAccessory.updateCurrentHeatingCoolingState(
          status.isSpaHeating
            ? this.Characteristic.CurrentHeatingCoolingState.HEAT
            : this.Characteristic.CurrentHeatingCoolingState.OFF,
        )

        this.spaThermostatAccessory.updateTargetHeatingCoolingState(
          this.mapHeatModeToTargetHeatingCoolingState(status.spaHeatMode),
        )
      }

      for (const circuitAccessory of this.circuitAccessories) {
        circuitAccessory.updateOn(
          status.circuitState.get(circuitAccessory.context.id) ? true : false,
        )
      }
    }
  }

  /** refresh if cached values are "stale" */
  private refreshIfNeeded(): boolean {
    const now = Date.now()
    if (now - this.lastRefreshTime > 15 * 1000) {
      // set now, so we don't trigger multiple...
      this.lastRefreshTime = now
      this.refreshStatus().catch(err => {
        this.log.error('refreshIfNeeded', err)
      })
      return true
    } else {
      return false
    }
  }

  /** convenience function to add an `on('get')` handler which refreshes accessory values if needed  */
  public triggersRefreshIfNeded(service: Service, type: CharacteristicType): void {
    const characteristic = service.getCharacteristic(type)
    characteristic.on('get', (callback: CharacteristicGetCallback) => {
      // just return current cached value, and refresh if needed
      callback(null, characteristic.value)
      if (this.refreshIfNeeded()) {
        this.log.debug('triggered refresh on get', service.displayName, characteristic.displayName)
      }
    })
  }

  public setTargetTemperature(context: ThermostatAccessoryContext, temperature: number): void {
    if (this.poolConfig === undefined) {
      this.log.warn('setTargetTemperature failed: poolConfig is undefined')
      return
    }
    // need to convert from Celsius to what pool is conifigured for
    const heatPoint = this.poolConfig.isCelsius ? temperature : Math.round(temperature * 1.8 + 32)
    this.controller
      .setHeatPoint(context.bodyType, heatPoint)
      .then(() => {
        this.log.debug(
          'setTargetTemperature: successfully set target temperature: ',
          context.bodyType,
          heatPoint,
        )
      })
      .catch(err => {
        this.log.error('setTargetTemperature', err)
      })
  }

  public setTargetHeatingCoolingState(context: ThermostatAccessoryContext, state: number): void {
    this.controller
      .setHeatMode(context.bodyType, this.mapTargetHeatingCoolingStateToHeatMode(state))
      .then(() => {
        this.log.debug(
          'setTargetHeatingCoolingState: successfully set target heating/cooling state: ',
          context.bodyType,
          state,
        )
      })
      .catch(err => {
        this.log.error('setTargetHeatingCoolingState', err)
      })
  }

  public setCircuitState(context: CircuitAccessoryContext, state: boolean): void {
    this.controller
      .setCircuitState(context.id, state)
      .then(() => {
        this.log.debug('setCircuitState: successfully set circuit state: ', context.id, state)
      })
      .catch(err => {
        this.log.error('setCircuitState', err)
      })
  }

  public sendLightCommand(_context: SetColorAccessoryContext, cmd: number): void {
    this.controller
      .sendLightCommand(cmd)
      .then(() => {
        this.log.debug('sendLightCommand: successfully sent light command: ', cmd)
        // sending the light command will turn on pool/spa lights, so refresh
        setTimeout(() => {
          this.refreshStatus().catch(err => {
            this.log.error('sendLightCommand refresh', err)
          })
        }, 2500)
      })
      .catch(err => {
        this.log.error('sendLightCommand', err)
      })
  }

  public generateUUID(accessorySalt: string) {
    if (this.poolConfig !== undefined) {
      return this.api.hap.uuid.generate(this.poolConfig.deviceId + ':' + accessorySalt)
    } else {
      this.log.error('poolConfig is undefined')
      return ''
    }
  }

  private backoff(retryAttempt: number, maxTime: number): number {
    retryAttempt = Math.max(retryAttempt, 1)
    return Math.min(Math.pow(retryAttempt - 1, 2) + Math.random(), maxTime)
  }

  public accessoryInfo(): {
    manufacturer: string
    model: string
    serialNumber: string
  } {
    if (this.poolConfig) {
      return {
        manufacturer: 'Pentair',
        // store software version in model, since it doesn't follow
        // proper n.n.n format Apple requires and model is a string
        model: this.poolConfig.softwareVersion,
        serialNumber: this.poolConfig.deviceId,
      }
    } else {
      this.log.error('poolConfig is null getting accessoryInfo')
      return {
        manufacturer: 'unknown',
        model: 'unknown',
        serialNumber: '',
      }
    }
  }

  private applyConfigDefaults(config: PlatformConfig) {
    // config.ip_address
    config.port = config.port ?? 80
    // config.username
    // config.password
    // config.hidden_circuits
    config.hideAirTemperatureSensor = config.hideAirTemperatureSensor ?? false
    config.hidePoolTemperatureSensor = config.hidePoolTemperatureSensor ?? false
    config.hideSpaTemperatureSensor = config.hideSpaTemperatureSensor ?? false
    config.hidePoolThermostat = config.hidePoolThermostat ?? false
    config.hideSpaThermostat = config.hideSpaThermostat ?? false
    config.statusPollingSeconds = config.statusPollingSeconds ?? 60
    config.createLightColorSwitches = config.createLightColorSwitches ?? false
    config.disabledLightColors = config.disabledLightColors ?? []
    this.log.debug('config', this.config)
  }

  /** normalize temperature to celsius for homekit */
  normalizeTemperature(temperature: number): number {
    return this.poolConfig?.isCelsius ? temperature : (temperature - 32) / 1.8
  }

  /** map pool heat mode to thermostat target heating/coooling state  */
  mapHeatModeToTargetHeatingCoolingState(poolHeatMode: number) {
    switch (poolHeatMode) {
      case Controller.HEAT_MODE_OFF:
        return this.Characteristic.TargetHeatingCoolingState.OFF
      case Controller.HEAT_MODE_HEAT_PUMP:
        return this.Characteristic.TargetHeatingCoolingState.HEAT
      case Controller.HEAT_MODE_SOLAR_PREFERRED:
        return this.Characteristic.TargetHeatingCoolingState.AUTO
      case Controller.HEAT_MODE_SOLAR:
        return this.Characteristic.TargetHeatingCoolingState.COOL
      default:
        return this.Characteristic.TargetHeatingCoolingState.OFF
    }
  }

  /** map thermostat target heating/coooling state to pool heat mode */
  mapTargetHeatingCoolingStateToHeatMode(targetHeatingCoolingState: number) {
    switch (targetHeatingCoolingState) {
      case this.Characteristic.TargetHeatingCoolingState.OFF:
        return Controller.HEAT_MODE_OFF
      case this.Characteristic.TargetHeatingCoolingState.HEAT:
        return Controller.HEAT_MODE_HEAT_PUMP
      case this.Characteristic.TargetHeatingCoolingState.AUTO:
        return Controller.HEAT_MODE_SOLAR_PREFERRED
      case this.Characteristic.TargetHeatingCoolingState.COOL:
        return Controller.HEAT_MODE_SOLAR
      default:
        return Controller.HEAT_MODE_UNCHANGED
    }
  }
}
